### ISSUE TITLE
Limit domain for google login, so it doesn't conflict with a subdomain?

### ISSUE THREAD

**Issue by BobStein** [authomatic/authomatic#218]
I have a pair of sites that both use authomatic with Flask-login for google OAuth2,

https://unslumping.org/
https://fun.unslumping.org/

Each works fine if I clear all cookies in both domains before I log in. But logging in to the 2nd level domain seems to mess up logging in to the 3rd level domain. I get caught in a loop where `.login()` keeps returning an object with a `.error` either "Unable to retrieve stored state!" or "The returned state csrf cookie ... doesn't match with the stored state!"

Is there a way I can limit the scope of cookies to the root domain, and not let them be used by the subdomain?

**Comment by mrichar1** [comment]
Hi - thanks for raising this issue.

I've just been having a look at the spec for setting cookies, which seems to be in RFC2109. It looks like the default position is that cookies shouldn't be shared, unless the `Set-Cookie` header has `Domain` set (which is not the default in `authomatic`). However if you set it o e.g. `example.com` then it is valid for all subdomains as well, so the only ay the `Domain` can be useful is if you have 2 siblings to separate, e.g. `www.example.com` and `foo.example.com`

Apparently one way round this is to use the Content Security Policy sandbox option, by setting the following header on all requests https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/sandbox

So you likely want to set a header such as:
```
Content-Security-Policy: sandbox allow-forms allow-scripts;
```

This should prevent the sharing of cookies between pages, and thus prevent this problem.

I've not tried this on Flask, but their docs suggest that you should be able to set a header on every response by doing:

```
@app.after_request
def apply_csp(response):
    response.headers["Content-Security-Policy"] = "sandbox allow-forms allow-scripts;"
    return response
```

Please let us know if this fixes your problem!

**Comment by vinay-kumar-3574** [comment]
@BobStein @mrichar1 can i work on this ?

**Comment by mrichar1** [comment]
@vinay-kumar-3574 yes this is an old issue with no active development so you're welcome to have a look at submitting a PR fix if you are still seeing this issue. Note that here has been significant development to Authomatic since this was reported, so it is worth testing against the most recent code to check the bug (and suggested solution) still apply.

**Comment by vinay-kumar-3574** [comment]
@mrichar1 
Root cause analysis

I investigated this on latest code and the problem is caused by **Flask session cookie collisions** if running more than one Flask app on the same parent domain (e.g. `unslumping.org` and `fun.unslumping.org`).

Authomatic stores OAuth CSRF/state values in the Flask `session`. If session cookies are shared across subdomains (for example via `SESSION_COOKIE_DOMAIN=.unslumping.org`), one app can overwrite the stored state of the other. When the OAuth redirect returns, the app reads a different CSRF value than it originally stored, resulting in errors such as:

- Unable to retrieve stored state

`- `The returned state csrf cookie doesn't match with the stored state`

Regarding CSP sandbox

CSP sandbox headers do not solve the root problem. Cookie sharing is controlled via `Domain` attributes on cookies, not via CSP. Using CSP sandbox in this way is unrelated to cookie scoping and could interfere with OAuth redirect flows.

The recommended solution is

That can be fixed by **isolating Flask session cookies per application**, for instance:
The code above can be copied and run as follows in iPython:
app.config['SESSION_COOKIE_NAME'] = 'unique_name_per_app' app.config['SESSION_COOKIE_DOMAIN'] = 'fun.unslumping.org'  # no leading dot

**Comment by vinay-kumar-3574** [comment]
@mrichar1 If this looks reasonable, I can submit a small documentation PR describing this configuration

**Comment by mrichar1** [comment]
Many thanks for identifying the issue and confirming it's coming via Flask and not an Authomatic issue. Yes I'd be happy for a docfix explaining this, which would let us close this issue.

**Comment by vinay-kumar-3574** [comment]
@mrichar1 I’ve opened a documentation PR that explains the Flask session configuration to avoid this issue: #249

**Comment by vinay-kumar-3574** [comment]
@mrichar1 ’m interested in contributing more regularly and would love to get involved further. Please let me know if there are areas where help is needed.

**Comment by mrichar1** [comment]
That would be great - we're always happy to have more contributors! @gnuamua has recently undertaken a big overhaul of the codebase so is probably in the best position to suggest areas where help would be useful.
